src/journal/compress.c

```diff
@@ -667,13 +677,237 @@ int decompress_stream_lz4(int in, int out, uint64_t max_bytes) {
+        _cleanup_free_ void *buffOut = NULL;
+        size_t buffInSize, buffOutSize;
+        size_t z;
+        uint64_t left = max_bytes, in_bytes = 0;
+        uint64_t inputSize = 0;
+
+        assert(fdf >= 0);
+        assert(fdt >= 0);
+
+        /* Create the context and buffers */
+        buffInSize = ZSTD_CStreamInSize();
+        buffOutSize = ZSTD_CStreamOutSize();
+        buffIn = malloc(buffInSize);
+        buffOut = malloc(buffOutSize);
+        cctx = ZSTD_createCCtx();
+        if (!cctx || !buffOut || !buffIn)
+                return -ENOMEM;
+
+        if (inputSize) {
+                z = ZSTD_CCtx_setPledgedSrcSize(cctx, inputSize);
+                if (z)
+                        log_debug("Failed to enable Zstd Inputsize: %s",
+                                ZSTD_getErrorName(z));
+        }
+
+        z = ZSTD_CCtx_setParameter(cctx, ZSTD_c_checksumFlag, 1);
+        if (z)
+                log_debug("Failed to enable Zstd Checksum: %s",
+                          ZSTD_getErrorName(z));
+
+        /* This loop read from the input file, compresses that entire chunk,
+         * and writes all output produced to the output file.
+         */
+        for (bool lastChunk = false; !lastChunk;) {
+                ZSTD_inBuffer input = {.src = buffIn, .size = 0, .pos = 0};
+                ssize_t red;
+
+                red = loop_read(fdf, buffIn, buffInSize, 1);
+                if (_unlikely_(red <= 0)) {
+                        if (red < 0)
+                                return -errno;
+                        lastChunk = true;
+                }
+
+                in_bytes += (size_t)red;
+                input.size = (size_t)red;
+
+                for (bool finished = false; !finished;) {
+                        ZSTD_outBuffer output = {
+                            .dst = buffOut, .size = buffOutSize, .pos = 0};
+                        size_t remaining;
+                        ssize_t wrote;
+
+                        /* Compress into the output buffer and write all of the
+                         * output to the file so we can reuse the buffer next
+                         * iteration.
+                         */
+                        remaining = ZSTD_compressStream2(
+                            cctx, &output, &input,
+                            lastChunk ? ZSTD_e_end : ZSTD_e_continue);
+
+                        if (_unlikely_(ZSTD_isError(remaining))) {
+                                log_debug("ZSTD encoder failed: %s",
+                                          ZSTD_getErrorName(remaining));
+                                return zstd_ret_to_errno(remaining);
+                        }
+
+                        if (_unlikely_(left < output.pos))
+                                return -EFBIG;
+
+                        wrote = loop_write(fdt, output.dst, output.pos, 1);
+                        if (_unlikely_(wrote < 0))
+                                return -errno;
```

Comment:
error is returned in return value, hence `return wrote;`

Reply:
ditto for `loop_read`